tt_metal/hw/inc/dataflow_api.h

```diff
@@ -1536,12 +1536,92 @@ enum class BufferType: uint8_t {
     SYSTEM_MEMORY = 2
 };
 
+FORCE_INLINE
+uint32_t min(uint32_t a, uint32_t b) { return (a < b) ? a: b; }
+
+
+template<bool READ>
+FORCE_INLINE void noc_async_sharded_read_write_helper(
+                                    const uint32_t num_cores,
+                                    const uint32_t page_size,
+                                    const uint32_t bank_base_address,
+                                    volatile tt_l1_ptr uint32_t* base_command_addr,
+                                    const uint32_t addr,
+                                    const uint32_t num_pages,
+                                    const uint32_t page_id
+                                    ){
+
+    uint32_t pages_start = 0;
+    uint32_t pages_end = 0;
+
+
+    //first get to correct core
+    uint32_t core_word_id = 0;
+    while (not (page_id >= pages_start and page_id < pages_end)) {
+        uint32_t num_pages_core = base_command_addr[core_word_id];
+        pages_end = pages_start + num_pages_core;
+        uint32_t core_id_x = base_command_addr[core_word_id + 1];
+        uint32_t core_id_y = base_command_addr[core_word_id + 2];
+        if (not (page_id >= pages_start and page_id < pages_end)) {
+            pages_start = pages_end;
+        }
+        core_word_id+=NUM_ENTRIES_PER_SHARD;
+    }
+
+    core_word_id-= NUM_ENTRIES_PER_SHARD;
+
+    uint32_t flattened_page_id = page_id;
+
+    uint32_t host_page_id = 0;
+    uint32_t host_offset = 0;
+    uint32_t core_page_id = (flattened_page_id - pages_start);
+    uint32_t core_offset = core_page_id * page_size;
+
+    uint32_t num_pages_left = num_pages;
+
+    while (num_pages_left > 0) {
+        uint32_t num_pages_core = base_command_addr[core_word_id];
+        pages_end = pages_start + num_pages_core;
+        uint32_t core_id_x = base_command_addr[core_word_id + 1];
+        uint32_t core_id_y = base_command_addr[core_word_id + 2];
+
+
+        //now curr_page_id pointing to beginning of section we want in this core
+        uint32_t num_pages_write_core = min(pages_end - flattened_page_id, num_pages_left);
+
+        uint32_t size_in_bytes_written = num_pages_write_core *page_size;
+
+        //Writing at beginning of core
+        uint64_t noc_address = get_noc_addr(core_id_x, core_id_y,
+                                        bank_base_address + core_offset);
+
+        if constexpr (READ) {
+            noc_async_read(noc_address, addr + host_offset, size_in_bytes_written);
+        }
+        else{
+            noc_async_write(addr + host_offset, noc_address, size_in_bytes_written);
+        }
+
+        num_pages_left-= num_pages_write_core;
+        host_offset += size_in_bytes_written;
+        core_offset = 0;
+        pages_start = pages_end;
+        flattened_page_id = pages_start;
+        core_word_id += NUM_ENTRIES_PER_SHARD;
+    }
+}
+
 class Buffer {
```

Comment:
have you considered creating a separate `ShardedBuffer` type that contains a `Buffer` object + the extra sharded fields?

Reply:
this would also remove need of separate `init` and `init_sharded` functions